lib/extract-js.js

```diff
@@ -0,0 +1,216 @@
+                    end: scriptEndPos
+                });
+            }
+
+            // reset script start position
+            scriptStartPos = 0;
+        }
+    }
+
+    var scriptStartPos = 0;
+    var scripts = [];
+    var parser = new htmlparser.Parser({
+        onopentag: onopen,
+        onclosetag: onclose
+    });
+
+    parser.parseComplete(html);
+
+    return scripts;
+}
+
+/**
+ * Trim script source and normalize lines offset.
+ * @param {String} source
+ * @returns {Object[]}
+ */
+function normalizeSource(source) {
+    var lines = source.split(/\r\n?|\n\r?/);
+    var lineCount = lines.length;
+    var lastLine = lines.pop();
+    var tabOnlyOffset = false;
+    var spaceOnlyOffset = false;
+    var offset;
+
+    // remove first list if it's an empty string
+    // usualy <script> starts with new line
+    if (!lines[0]) {
+        lines.shift();
+    }
+
+    // replace last line by empty string if it contains only whitespaces
+    // it helps avoid disallowTrailingWhitespace errors on last line
+    lines.push(/\S/.test(lastLine) ? lastLine : '');
+
+    // calculate min line offset
+    offset = Math.min.apply(null, lines.map(function(line) {
+        // skip empty lines
+        if (!line) {
+            return Infinity;
+        }
+
+        // fetch whitespaces at the line begining
+        var offsetStr = line.match(/^\s*/)[0];
+        var tabCount = offsetStr.match(/\t*/)[0].length;
+
+        // mixed spaces and tabs in one offset -> don't remove offsets
+        if (tabCount && tabCount !== offsetStr.length) {
+            return 0;
+        }
+
+        if (tabCount) {
+            if (spaceOnlyOffset) {
+                // no spaces, but previous offset has ony spaces -> mixed spaces and tabs
+                return 0;
+            } else {
+                // remember offset contains only tabs
+                tabOnlyOffset = true;
+            }
+        } else {
+            if (tabOnlyOffset) {
+                // no tabs, but previous offset has only tabs -> mixed spaces and tabs
+                return 0;
+            } else {
+                // remember offset contains only spaces
+                spaceOnlyOffset = true;
+            }
+        }
+
+        return offsetStr.length;
+    }));
+
+    // remove common offsets if possible
+    if (offset) {
+        lines = lines.map(function(line) {
+            return line.substr(offset);
+        });
+    }
+
+    return {
+        source: lines.join('\n'),
+        offset: offset,
+        lineCount: lineCount
+    };
+}
+
+/**
+ * Parse HTML and search for <script> sources. Each script source also normalize
+ * by line offset. Result contains script sources with information about line
+ * offset (that was removed for each line) and lines count before script source.
+ * This information helps restore absolute positions in html file for errors.
+ * @param {String}
+ * @returns {Object[]}
+ */
+function extractJs(filename, data) {
```

Comment:
Consider renaming `data` to `source` as it's more descriptive and matches the constructor's signature.